src/storage/data_table.cpp

```diff
@@ -59,6 +61,54 @@ void DataTable::Scan(const common::ManagedPointer<transaction::TransactionContex
   out_buffer->SetNumTuples(filled);
 }
 
+void DataTable::NUMAScan(common::ManagedPointer<transaction::TransactionContext> txn,
+                         std::vector<ProjectedColumns *> *out_buffers, ProjectedColumns *const result_buffer,
+                         common::ExecutionThreadPool *thread_pool) {
+  std::vector<numa_region_t> numa_regions;
+  GetNUMARegions(&numa_regions);
+
+  // TODO(emmanuel) add additional buffers to the array so that the assert below is unnecessary
+
+  TERRIER_ASSERT(out_buffers->size() >= numa_regions.size(),
+                 "should have at least as many outbuffers as numa regions currently occupied");
+  for (uint32_t i = 0; i < numa_regions.size(); i++) {
+    // Lambda function to pass into thread pool
+    ProjectedColumns *out_buffer = (*out_buffers)[i];
+    numa_region_t region = numa_regions[i];
+    auto lambda = [&, out_buffer, region] {
+      uint32_t filled = 0;
+
+      for (auto it = begin(region); it != end(region) && filled < out_buffer->MaxTuples(); it++) {
+        ProjectedColumns::RowView row = out_buffer->InterpretAsRow(filled);
+        const TupleSlot slot = *it;
+        // Only fill the buffer with valid, visible tuples
+        if (SelectIntoBuffer(txn, slot, &row)) {
+          out_buffer->TupleSlots()[filled] = slot;
+          filled++;
+        }
+      }
+      out_buffer->SetNumTuples(filled);
+    };
+
+    if (thread_pool == nullptr) {
+      lambda();
+      continue;
+    }
+    std::promise<void> p;
+    thread_pool->SubmitTask(&p, lambda, region);
+    p.get_future().get();
+  }
+
+  uint32_t result_index = 0;
+  for (auto cur_buffer : *out_buffers) {
```

Comment:
I wonder if this would become a bottleneck for low selectivity... would be nice if these threads coordinate with each other and write to result_buffer concurrently.

Reply:
This is pretty much a function we just use for testing that will probably be removed once you guys add in the parallel scan functionality.

Reply:
We no longer need this function so it has been deleted